### PR TITLE
TST add test for dtype consistency with pd replace #23305

### LINKED ISSUE #23305: Columns lose category dtype after calling replace on the dataframe
xref https://github.com/pandas-dev/pandas/pull/25521

#### Code Sample, a copy-pastable example if possible

```python
>>> import pandas as pd
>>> pd.__version__
'0.23.4'

# Sample dataframe with two categorical and one int column
>>> df = pd.DataFrame(
...     [[1., 'A', 'x'], [2, 'B', 'y'], [3, 'C', 'z']],
...     columns=['first', 'second', 'third']
... ).astype({'second': 'category', 'third': 'category'})

# Replace int values
>>> df = df.replace(1, 10)

# Both categorical columns turned into object...
>>> df.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
first     3 non-null float64
second    3 non-null object
third     3 non-null object
dtypes: float64(1), object(2)
memory usage: 152.0+ bytes

# Sample dataframe with two categorical columns
>>> df = pd.DataFrame(
...     [['A', 'x'], ['B', 'y'], ['C', 'z']],
...     columns=['first', 'second',]
... ).astype({'first': 'category', 'second': 'category'})

>>> df.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
first     3 non-null category
second    3 non-null category
dtypes: category(2)
memory usage: 294.0 bytes

# Replace values in column `first`
>>> df = df.replace('A', 'B')

# Dtype of column `second` becomes object
>>> df.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
first     3 non-null category
second    3 non-null object
dtypes: category(1), object(1)
memory usage: 211.0+ bytes
```
#### Problem description

Calling `replace()` on a dataframe changes `category` columns' dtype to `object` in an apparently inconsisent manner:

1. When `replace()` contains non-categorical values that are not present in either categorical column, all categorical dtype turn into `object`.
2. When `replace()` contains categories from a `category` dtype column, then that column keeps its dtype, but other categorical columns turn into `object`.

See the examples above for both. Even if it is somehow intentional I find it quite confusing.

#### Expected Output

I would expect the categorical columns to keep their `category` dtype after `replace` is called on the dataframe (at least for those categorical columns that are unaffected by `replace`).

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None

pandas: 0.23.4
pytest: None
pip: 18.0
setuptools: 39.0.1
Cython: None
numpy: 1.15.0
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by mathurk1**
- [x] closes #23305 
- [x] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/methods/test_replace.py [hunk 1]
```diff
@@ -1420,3 +1478,31 @@ def test_replace_period_ignore_float(self):
         result = df.replace(1.0, 0.0)
         expected = pd.DataFrame({"Per": [pd.Period("2020-01")] * 3})
         tm.assert_frame_equal(expected, result)
+
+    def test_replace_value_category_type(self, input_category_df, expected_category_df):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:1482]: i would just inline the source and expected frame as you are only using them *inside* this test

File: pandas/tests/frame/methods/test_replace.py [hunk 2]
```diff
@@ -1420,3 +1478,31 @@ def test_replace_period_ignore_float(self):
...
+
+        tm.assert_frame_equal(input_df, expected_category_df)
+
+    @pytest.mark.xfail(
+        reason="category dtype gets changed to object type after replace, see #23305",
+        strict=True,
+    )
+    def test_replace_dict_category_type(self, input_category_df, expected_category_df):
+        """
+        Test to ensure category dtypes are maintained
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:1501]: just make a simple example that you can inline here, no need for the fixtures. Please create a new issue for this case (as we will close the original with this PR). indicate the issue number in the xfail

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/methods/test_replace.py (hunk 2 above)
Comment (mathurk1) [reply pandas/tests/frame/methods/test_replace.py:1481]: I have raised a new issue for the bug - https://github.com/pandas-dev/pandas/issues/35268 and referenced it in the xfail.

Comment (mathurk1) [reply pandas/tests/frame/methods/test_replace.py:1481]: Hi @jreback, please let me know if we need anything else to merge this PR. Thanks!

File: pandas/tests/frame/methods/test_replace.py [hunk 3]
```diff
@@ -1420,3 +1420,83 @@ def test_replace_period_ignore_float(self):
...
+
+        # create expected dataframe
+        expected_dict = {"col1": ["z"], "col2": ["obj9"], "col3": ["catX"]}
+        # explicitly cast columns as category
+        expected_df = pd.DataFrame(data=expected_dict).astype(
+            {"col1": "category", "col2": "category", "col3": "category"}
+        )
+
+        # replace values in input dataframe using a dict
+        input_df = input_df.replace({"a": "z", "obj1": "obj9", "cat1": "catX"})
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:1500]: can u call this result

File: pandas/tests/frame/methods/test_replace.py [hunk 4]
```diff
@@ -1420,3 +1420,83 @@ def test_replace_period_ignore_float(self):
...
+        input_dict = {"col1": ["a"], "col2": ["obj1"], "col3": ["cat1"]}
+        # explicitly cast columns as category
+        input_df = pd.DataFrame(data=input_dict).astype(
+            {"col1": "category", "col2": "category", "col3": "category"}
+        )
+
+        # create expected dataframe
+        expected_dict = {"col1": ["z"], "col2": ["obj9"], "col3": ["catX"]}
+        # explicitly cast columns as category
+        expected_df = pd.DataFrame(data=expected_dict).astype(
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:1495]: can u call this expected

File: pandas/tests/frame/methods/test_replace.py [hunk 5]
```diff
@@ -1420,3 +1420,83 @@ def test_replace_period_ignore_float(self):
...
+        expected_df["col4"] = expected_df["col4"].cat.reorder_categories(
+            ["cat1", "catX", "cat3", "cat4"], ordered=True
+        )
+
+        # replace values in input dataframe
+        input_df = input_df.replace("d", "z")
+        input_df = input_df.replace("obj1", "obj9")
+        input_df = input_df.replace("cat2", "catX")
+
+        tm.assert_frame_equal(input_df, expected_df)
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:1473]: similar to below
call this result and expected

File: pandas/tests/frame/methods/test_replace.py [hunk 6]
```diff
@@ -1420,3 +1420,83 @@ def test_replace_period_ignore_float(self):
         result = df.replace(1.0, 0.0)
         expected = pd.DataFrame({"Per": [pd.Period("2020-01")] * 3})
         tm.assert_frame_equal(expected, result)
+
+    def test_replace_value_category_type(self):
+        """
+        Test to ensure category dtypes are maintained
+        after replace with direct values
+        """
```
Comment (jreback) [inline pandas/tests/frame/methods/test_replace.py:1428]: can u add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
looks good some comments; ping on green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mathurk1

**Pull request merged**